### DISCUSSION TITLE
When Creating Named Class Nodes & Inheritance

### DISCUSSION THREAD

**Discussion by Paellu** [godotengine/godot-proposals#5753] [Editor]
# Problem
When creating a new node from the scene dock that uses `class_name` that script gets attached which is not the expected behavior.
So in essence you're creating the same node not a new one since built-in nodes are presented differently.

## Built-in Node behavior
Create a new node (example **Label**) and attach a script.

```
extends Label

# Called when the node enters the scene tree for the first time.
func _ready():
	pass # Replace with function body.


# Called every frame. 'delta' is the elapsed time since the previous frame.
func _process(delta):
	pass
```

# Additional Problems
The standard behavior also presents additional problems when say for instance an add-on creates named class nodes
All the code gets exposed when it should extend its functionality like the above code.

Creating Inheritance becomes a nightmare since you can accidentally delete the original `named_class` 
code.

# Solutions

When a new `named class node` is created the script attached could be a new built-in script
```
extends named_class

# Called when the node enters the scene tree for the first time.
func _ready():
	pass # Replace with function body.


# Called every frame. 'delta' is the elapsed time since the previous frame.
func _process(delta):
	pass
```
that the user can decide to save externally if desired later

or even better when a new `named class node` is selected the attach script dialog can open after with the correct inheritance and the user can decide what to do instantly

Example

Create New Node             |  Attach Script
:-------------------------:|:-------------------------:
![](https://imgur.com/3gmEFOx.png)  |  ![](https://imgur.com/AdHjfM8.png)

And as seen in Godot 4 if the script has `@tool` always show the description of the `named_class` or otherwise the actual class contents
<img src="https://imgur.com/M7r0LxI.png" alt="Logo"/>

**Comment by dalexeev** [comment]
> Creating Inheritance becomes a nightmare since you can accidentally delete the original named_class
code.

See https://github.com/godotengine/godot-proposals/issues/4890#issuecomment-1186428851.
